drivers/include/onewire.h

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright (C) 2017 Frits Kuipers
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @defgroup    drivers_ds18 DS18 temperature sensor driver
+ * @ingroup     drivers_sensors
+ * @ingroup     drivers_saul
+ * @brief       Driver interface for the DS18 temperature sensors
+ *
+ * This driver provides @ref drivers_saul capabilities.
+ * Currently the driver has the following limitations:
+ *- Does not allow addressing devices, only supports a single device on the bus.
+ *- The 1-Wire bus handling is hardcoded to the driver.
+ *- Does not allow configuration of sampling width.
+ *
+ * @note Due to timing issues present on some boards this drivers features two
+ * ways of reading information from the sensor. The optimized uses accurate
+ * delays to handle this, while the second way polls the line for changes. If
+ * you know that your board can handle ~3us resolution with the xtimer module,
+ * then the optimized way is recommended. To used the optimized way add the
+ * ds18_optimized module. Also this driver test application has a whitelist of
+ * the boards this driver has been tested on and known to work.
+ *
+ * @{
+ *
+ * @file
+ * @brief       Driver for Maxim Integrated DS1822 and DS18B20 temperature
+ *              sensors.
+ *
+ * @author      Frits Kuipers <frits.kuipers@gmail.com>
+ */
+
+#ifndef ONEWIRE_H
+#define ONEWIRE_H
+
+#include <stddef.h>
+#include <stdint.h>
+
+#include "mutex.h"
+#include "periph/gpio.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @name ds18 commands
+ * @{
+ */
+#define DS18_CMD_RSCRATCHPAD        (0xbe)
+#define DS18_CMD_WRITESCRATCHPAD    (0x4e)
+#define DS18_CMD_COPYSCRATCHPAD     (0x48)
+#define DS18_CMD_RECALLE            (0xb8)
+#define DS18_CMD_RPWRSPPLY          (0xb4)
+#define DS18_CMD_SEARCHROM          (0xf0)
+
+#define ONEWIRE_CMD_READROM            (0x33)
+#define ONEWIRE_CMD_MATCHROM           (0x55)
+#define ONEWIRE_CMD_SEARCHROM          (0xf0)
+#define ONEWIRE_CMD_ALARMSEARCH        (0xec)
+#define ONEWIRE_CMD_SKIPROM            (0xcc)
+/** @} */
+
+/**
+ * @brief Device initialization parameters
+ */
+typedef struct {
+    gpio_t pin;             /**< Pin the sensor is connected to */
+    gpio_mode_t out_mode;   /**< Pin output mode */
+    gpio_mode_t in_mode;    /**< Pin input mode (usually deduced from output mode) */
```

Comment:
?? never set but used ??

i remember to have seen this in #14897 and there i wrote: "if it is deduced from omode there is no need to save it"

https://github.com/RIOT-OS/RIOT/pull/14897#issuecomment-697386642

Reply:
I used a copy of the ds18 as my starting point. I agree this is not as it should be. Up until now, I've mostly concerned myself with the API and not the implementation, since I knew the implementation was at least mature. I'll make sure this gets cleaned up.